File: pages/python/src/pages/patch.py
"""Deterministic patch engine for Copilot Pages in Python.

Directly remedies RegexParseException and connection lockouts from DEF-PAGES-01.
"""

from typing import Tuple, Optional
from .ast import find_section
from .types import Page, EditResult


def replace_literal(
    content: str,
    target_text: str,
    replacement_text: str,
    occurrence_index: int = 0,
) -> Tuple[bool, str, Optional[int], Optional[str]]:
    """Replace an exact character-sequence target without regular expressions."""
    if not target_text:
        return False, content, None, "Target text cannot be empty."

    indices = []
    start = 0
    while True:
        pos = content.find(target_text, start)
        if pos == -1:
            break
        indices.append(pos)
        start = pos + len(target_text)

    if not indices:
        return False, content, None, f"Target text not found in document: '{target_text[:40]}...'"

    if occurrence_index >= len(indices):
        return (
            False,
            content,
            None,
            f"Occurrence index {occurrence_index} exceeds matches found ({len(indices)}).",
        )

    target_idx = indices[occurrence_index]
    new_content = (
        content[:target_idx] + replacement_text + content[target_idx + len(target_text):]
    )

    return True, new_content, target_idx, None


def replace_section(
    content: str,
    heading: str,
    new_section_content: str,
) -> Tuple[bool, str, Optional[str]]:
    """Replace a bounded section by heading without regex matching."""
    sec = find_section(content, heading)
    if not sec:
        return False, content, f"Section with heading '{heading}' not found."

    formatted_new_content = new_section_content
    if not formatted_new_content.endswith("\n") and sec.end_index < len(content):
        formatted_new_content += "\n"

    new_content = (
        content[:sec.start_index] + formatted_new_content + content[sec.end_index:]
    )

    return True, new_content, None


def apply_atomic_edit(
    page: Page,
    kind: str,
    target: str,
    replacement: str,
    occurrence_index: int = 0,
) -> EditResult:
    """Apply an atomic edit operation to a Page, updating its revision token."""
    if kind == "literal":
        success, new_content, idx, err = replace_literal(
            page.content, target, replacement, occurrence_index
        )
        summary = f"Replaced literal text at offset {idx}"
    elif kind == "section":
        success, new_content, err = replace_section(page.content, target, replacement)
        idx = None
        summary = f"Replaced section '{target}'"
    else:
        return EditResult(success=False, content=page.content, error=f"Unknown edit kind '{kind}'")

    if not success:
        return EditResult(success=False, content=page.content, error=err)

    import time
    now_ts = str(time.time())
    new_rev = f"rev-py-{abs(hash(new_content))}-{int(time.time())}"

    updated_page = Page(
        id=page.id,
        title=page.title,
        content=new_content,
        revision_id=new_rev,
        updated_at=now_ts,
        is_canonical=page.is_canonical,
        canonical_of=page.canonical_of,
        metadata=page.metadata,
    )

    return EditResult(
        success=True,
        content=new_content,
        page=updated_page,
        applied_change_summary=summary,
        match_index=idx,
    )

File: pages/lib/src/patch.ts
import { findSection } from "./parser.js";
import { Page, EditOperation, EditResult, LiteralEditOp, SectionEditOp } from "./types.js";

/**
 * Replaces a literal text passage within content without using regular expressions.
 * Directly eliminates RegexParseException (DEF-PAGES-01).
 */
export function replaceLiteral(
  content: string,
  targetText: string,
  replacementText: string,
  occurrenceIndex: number = 0
): { success: boolean; content: string; matchIndex?: number; error?: string } {
  if (!targetText) {
    return { success: false, content, error: "Target text cannot be empty." };
  }

  const indices: number[] = [];
  let pos = content.indexOf(targetText);
  while (pos !== -1) {
    indices.push(pos);
    pos = content.indexOf(targetText, pos + targetText.length);
  }

  if (indices.length === 0) {
    return {
      success: false,
      content,
      error: `Target text not found in document: "${targetText.slice(0, 50)}${targetText.length > 50 ? "..." : ""}"`,
    };
  }

  if (occurrenceIndex >= indices.length) {
    return {
      success: false,
      content,
      error: `Requested occurrence index ${occurrenceIndex} exceeds matches found (${indices.length}).`,
    };
  }

  const targetIndex = indices[occurrenceIndex];
  const newContent =
    content.slice(0, targetIndex) +
    replacementText +
    content.slice(targetIndex + targetText.length);

  return {
    success: true,
    content: newContent,
    matchIndex: targetIndex,
  };
}

/**
 * Replaces a bounded section identified by its Markdown heading.
 * Bounds are computed structurally, guaranteeing no collateral modification to adjacent sections.
 */
export function replaceSection(
  content: string,
  heading: string,
  newSectionContent: string
): { success: boolean; content: string; error?: string } {
  const section = findSection(content, heading);
  if (!section) {
    return {
      success: false,
      content,
      error: `Section with heading "${heading}" not found.`,
    };
  }

  // Ensure trailing newline before next section starts if needed
  let formattedNewContent = newSectionContent;
  if (!formattedNewContent.endsWith("\n") && section.endIndex < content.length) {
    formattedNewContent += "\n";
  }

  const newContent =
    content.slice(0, section.startIndex) +
    formattedNewContent +
    content.slice(section.endIndex);

  return {
    success: true,
    content: newContent,
  };
}

/**
 * Generates a deterministic revision token for a page state.
 */
export function generateRevisionId(content: string, timestamp: string): string {
  let hash = 0;
  const str = `${timestamp}:${content}`;
  for (let i = 0; i < str.length; i++) {
    hash = (hash << 5) - hash + str.charCodeAt(i);
    hash |= 0;
  }
  return `rev-${Math.abs(hash).toString(16)}-${Date.now().toString(36)}`;
}

/**
 * Applies an atomic edit operation to a Page, updating its revision token and timestamp.
 */
export function applyAtomicEdit(page: Page, edit: EditOperation): EditResult {
  let patchResult: { success: boolean; content: string; error?: string; matchIndex?: number };

  if (edit.kind === "literal") {
    patchResult = replaceLiteral(
      page.content,
      edit.targetText,
      edit.replacementText,
      edit.occurrenceIndex ?? 0
    );
  } else if (edit.kind === "section") {
    patchResult = replaceSection(
      page.content,
      edit.heading,
      edit.newSectionContent
    );
  } else {
    return {
      success: false,
      error: `Unsupported edit operation kind: ${(edit as { kind: string }).kind}`,
    };
  }

  if (!patchResult.success) {
    return {
      success: false,
      error: patchResult.error,
    };
  }

  const now = new Date().toISOString();
  const updatedPage: Page = {
    ...page,
    content: patchResult.content,
    updatedAt: now,
    revisionId: generateRevisionId(patchResult.content, now),
  };

  const summary =
    edit.kind === "literal"
      ? `Replaced literal passage (${edit.targetText.length} chars) at offset ${patchResult.matchIndex}`
      : `Replaced section "${edit.heading}"`;

  return {
    success: true,
    page: updatedPage,
    appliedChangeSummary: summary,
    matchIndex: patchResult.matchIndex,
  };
}
